working/333 - shared memory multithreading/proposal.md

```diff
@@ -0,0 +1,561 @@
+
+```  dart
+/// Mark class sharable by implementing a [Shareable]
+/// interface.
+class S implements Shareable {
+  // ...
+} 
+```
+
+To maintain heap isolation we restrict the kind of data that a shareable class can contain: **Fields of sharable classes can only contain references to instances of sharable classes.** There is however no requirement of immutability imposed on shareable classes.
+
+References to shareable instances can be passed between isolates within an isolate group directly: for example sending a shareable instance through `SendPort` will not copy it.
+
+```dart  
+class S implements Shareable {
+  int v = 0;
+}
+
+final s = S();
+
+await Isolate.run(() {
+  // Even though we are running in a different isolate we have
+  // access to the original `s` and mutations performed in this
+  // isolate will be visible to other isolates.
+  // 
+  // (Subject to memory model constraints).
+  s.v = 10; 
+});
+
+expect(equals(10), s.v);
+```
+
+To make state sharing between isolates simpler we also allow to declare static fields which are shared between all isolates in the isolate group. `@shared` global fields are required to have static type that is a subtype of `Shareable?`. 
+
+```dart 
+// All isolates within an isolate group share this variable.
+@shared int v = 0;
+```
+
+### Functions
+
+Given the parallel between functions and other objects it makes sense to declare that functions which close only over shareable state are themselves shareable.
+
+**Static type of a function literal which captures only `Shareable` variables is  `ShareableFunction<F extends Function>` which is subtype of both `F` and `Shareable`.**
+
+```dart
+void main() {
+  final Shareable s = ...;
+  final f = () => s;  // ShareableFunction<Shareable Function()>
+
+  final Object o;
+  final g = () => o;  // Not shareable
+}
+
+class S implements Shareable {
+  void Function() f; // compile time error
+  
+  ShareableFunction<void Function()> f;  // ok
+}
+```
+
+Shareable functions have one subtle behavior - they capture and share local variables, but not the global state of the isolate. Consider for example the following code:
+
+``` dart
+int global = 0; 
+
+void main() async {
+  final s = S();
+  
+  s.v = 42;
+  global = 42;
+  
+  await Isolate.run(() {
+    print(global);  // => 0
+    global = 24;
+
+    print(s.v);  // 42
+    s.v = 24;
+  });
+  
+  print(global);  // 42
+  print(s.v);  // 24
+}
+```
+
+This behavior might be somewhat surprising to developers - especially if a functions depends on or updates some mutable global state. Whether this is a problem in practice is unclear: sending a closures through a `SendPort` leads to similar behavior.
```

Comment:
Instead of having surprising behavior, would it be possible to make it an error or lint? If it's not possible, consider explaining this in this paragraph.